Automattic/wp-calypso
# Jetpack Backup: Setup redux state for pre-flight checks

Resolves https://github.com/Automattic/jetpack-backup-team/issues/288 and partially https://github.com/Automattic/jetpack-backup-team/issues/289.

This PR introduces state management for restore preflight checks, introducing a `getPreflightStatus` selector that we could use for prompting users to introduce server credentials and a `usePreflightStatusQuery` hook that we could use in any component that requires to listen the preflight check status. The changes are foundational and currently have no direct user impact.

## Proposed Changes

* Setup redux state for pre-flight checks, it will be under `state.rewind[ siteId ].preflight`
* Add `getPreflightStatus` selector. It will return the overall status depending on the status of each individual test.
  - If any test has failed, the overall status is set to FAILED.
  - If no tests have failed but at least one is in progress, the overall status is IN_PROGRESS.
  - If all tests are successful, the overall status is SUCCESS.
  - If none of the above conditions are met, the overall status defaults to PENDING.
* Add `jetpack/backup-restore-preflight-checks` feature flag. If this flag is not enabled, the preflight tests' status returned by `getPrefllightStatus` selector will be FAILED by default, because we are planning to show the banner to enter credentials if the preflight tests fail.
* Add `usePreflightStatusQuery` hook that will be responsible for fetching the preflight status from `/sites/${ siteId }/rewind/preflight` every 5 seconds until a `SUCCESS` or `FAILED` overall status occurs.
* Add unit tests for reducer, selector and utils.

## Testing Instructions
This code will not impact any user at the moment, given we are just preparing all the redux setup needed to start managing preflight check status, so for the moment just ensure unit tests are passing:

```yarn run test-client client/state/rewind/preflight```

## Pre-merge Checklist
